### PR TITLE
Parameterized reuses TestClass

### PR DESCRIPTION
**Pull request by kcooney**

### CODE REVIEW

**Comment by kcooney** [comment]
@stefanbirkner could you review this?

**Comment by stefanbirkner** [comment]
I have a look.

File: src/main/java/org/junit/runners/Suite.java [hunk 1]
```diff
@@ -14,6 +14,7 @@
 import org.junit.runner.notification.RunNotifier;
 import org.junit.runners.model.InitializationError;
 import org.junit.runners.model.RunnerBuilder;
+import org.junit.runners.model.TestClass;
```
Comment (panchenko) [inline src/main/java/org/junit/runners/Suite.java]: This import seems not needed.

File: src/main/java/org/junit/internal/Checks.java [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+     * as the exception message if an exception is thrown.
+     *
+     * @param value object reference to check
+     * @param message message to use if {@code value} is {@code null}
+     * @return the passed-in value, if not {@code null}
+     * @throws NullPointerException if {@code value} is {@code null}
+     */
+    public static <T> T notNull(T value, String message) {
+        if (value == null) {
+            throw new NullPointerException(String.valueOf(message));
```
Comment (panchenko) [inline src/main/java/org/junit/internal/Checks.java]: I would suggest passing `message` directly without `String.valueOf()`.

Comment (kcooney) [inline src/main/java/org/junit/internal/Checks.java]: `message` can  be `null`

Comment (panchenko) [inline src/main/java/org/junit/internal/Checks.java]: In this case this method would behave exactly as the one above, similarly to `java.util.Objects.requireNonNull()`

File: src/main/java/org/junit/internal/builders/JUnit4Builder.java [hunk 3]
```diff
@@ -1,12 +1,12 @@
...
 
 public class JUnit4Builder extends RunnerBuilder {
     @Override
     public Runner runnerForClass(Class<?> testClass) throws Throwable {
-        return new BlockJUnit4ClassRunner(testClass);
+        return new JUnit4(testClass);
     }
-}
\ No newline at end of file
+}
```
Comment (stefanbirkner) [inline src/main/java/org/junit/internal/builders/JUnit4Builder.java]: This change is unrelated to the reuse of TestClass. It should be a separate commit.

File: src/main/java/org/junit/internal/Checks.java [hunk 4]
```diff
@@ -0,0 +1,37 @@
+package org.junit.internal;
+
+/** @since 4.13 */
+public final class Checks {
```
Comment (stefanbirkner) [inline src/main/java/org/junit/internal/Checks.java]: I think we should not extract a Checks class and simply inline the validations, in order to not provide methods someone may use and depend on. I know that is a very weak argument. I leave it up to you whether you want to keep it or not.

File: src/main/java/org/junit/internal/Checks.java (hunk 2 above)
Comment (stefanbirkner) [inline src/main/java/org/junit/internal/Checks.java]: I conform to @panchenko's suggestion. I normally expect that `someMethod(someValue)` and `someMethod(someValue, null)` behave the same.

File: src/main/java/org/junit/runners/Suite.java (hunk 1 above)
Comment (kcooney) [inline src/main/java/org/junit/runners/Suite.java]: Removed

File: src/main/java/org/junit/internal/Checks.java (hunk 2 above)
Comment (kcooney) [inline src/main/java/org/junit/internal/Checks.java]: The behavior I have here matches Guava's `Preconditions.checkNotNull(T, String)` method. I think the Guava team has given this much more thought than I have, but I can take a stab at explaining why the behavior I have is better.

IMHO it is a programmer error to pass in a `null` as the second parameter. If we have this method behave like `notNull(T)` when you pass in a `null` message we would be ignoring that programmer error. Having the message be `"null"` makes it clear that 1) someone tried to specify a real message, and 2) they messed up. Hopefully someone would notice and fix it.

File: src/main/java/org/junit/internal/builders/JUnit4Builder.java (hunk 3 above)
Comment (kcooney) [inline src/main/java/org/junit/internal/builders/JUnit4Builder.java]: It is a separate commit. See https://github.com/junit-team/junit4/pull/1449/commits/11822c4fd1d2388079ed6a2eadd61ed7b3d44035

I have this commit in this pull since there's no reason for `ParentRunner.createTestClass()` to be called for JUnit4-style tests that aren't annotated with `RunWith`.

File: src/main/java/org/junit/internal/Checks.java (hunk 4 above)
Comment (kcooney) [inline src/main/java/org/junit/internal/Checks.java]: As you can see in the commits, we have multiple classes that do `null` checks. Quite often when I write code in JUnit I find myself wanting a one-liner to check for null.

I doubt anyone would depend on this class, since 1) it's in an internal package, 2) it has an odd name, and 3) most everyone that is using JDK 6 or later would use `Objects.requireNotNull()` or Guava's `Preconditions.checkNotNull()`.

File: src/main/java/org/junit/internal/Checks.java (hunk 2 above)
Comment (panchenko) [inline src/main/java/org/junit/internal/Checks.java]: Yes, please, lets keep it simple. Also the 1.8 version with `Supplier<String>` passes the result directly to the NullPointerException constructor.

File: src/main/java/org/junit/internal/Checks.java (hunk 4 above)
Comment (stefanbirkner) [inline src/main/java/org/junit/internal/Checks.java]: I would like to go with the JDK way.

File: src/main/java/org/junit/internal/builders/JUnit4Builder.java (hunk 3 above)
Comment (stefanbirkner) [inline src/main/java/org/junit/internal/builders/JUnit4Builder.java]: Sorry, my fault. I was not aware that the PR is made up of multible commits.

File: src/main/java/org/junit/internal/Checks.java (hunk 2 above)
Comment (kcooney) [inline src/main/java/org/junit/internal/Checks.java]: Changed `notNull(T, String)` to not use `String.valueOf()` on the message.

**Comment by tasomaniac** [comment]
How does this work when we provide the parameters through constructor? In such cases, the properties in test instance are `private final`

**Comment by panchenko** [comment]
@tasomaniac This issue is about a _runner_ for parameterizzed tests - the tests are developed the same way as before.

**Comment by tasomaniac** [comment]
I see. The title says reuses TestClass. I thought TestClass is the instance of the test class we create.
